lib/kafka/producer.rb

```diff
@@ -357,6 +365,7 @@ def assign_partitions!
             exception: [e.class.to_s, e.message],
           })
 
+          topics_with_failures << message.topic
           failed_messages << message
```

Comment:
I don't see how this fixed the problem – you'd have to wait with writing the messages to `@buffer` until a partition has been assigned for each message in a topic.